### ISSUE TITLE
Dashboard 1.11.1 msg._client issues (!)

### ISSUE THREAD

**Issue by SynoUser-NL** [FlowFuse/node-red-dashboard#947]
### Current Behavior

With Dashboard V1.11.1, we are experiencing issues with regard to client data separation in Dashboard 2. In Dashboard 2 configuration, all Client Data options as well as FF Auth\ Client Constraints have been ticked.
Behaviour:
1) data is shared over browser sessions with _client enabled \ unaltered in the message
2) msg data sent to an input node is not displayed in the node within the same or other sessions
3) msg data leaving the input node and sent to a change node (to be stored in memory to form_name.formSubmissions[msg._client.user.username][msg.topic] ) results in an error if msg._client is not deleted first.
This is different from earlier behaviour (V1.10.x) where this worked fine for us.

### Expected Behavior

- sending a msg to an input node with msg._client displays the data sent in payload (this does not work if msg._client is not deleted!)
- storing a payload value with a change node with msg._client in the message should not result in an error.

### Steps To Reproduce

1) Enable client data & FF Auth \ client constraints for all nodes
2) connect DB2 input node & DB2 template node, configure template node to display payload.
3) connect and send input node output to change node, where payload is saved to form.std_vpn.formSubmissions[msg._client.user.username][msg.topic]
4) open dashboard in multiple browsers
5) in Dashboard 2 front-end, entering a message in the input box will be displayed in the all browser sessions (which should not happen)
6) the message to the change node will result in an error
7) to remove the error, place a change node before the change node and delete _client object
8) repeat 5, there is no error message

### Environment

- Dashboard version: 1.11.1
- Node-RED version: 3.19.x
- Node.js version: 18.x.x
- npm version: -
- Platform/OS: docker on linux (flowfuse)
- Browser: firefox,  opera, vivaldi, private sessions of all

### Have you provided an initial effort estimate for this issue?

I am not a FlowFuse team member

**Comment by Steve-Mcl** [comment]
_FYI: This is a duplicate of the reply on HS_


Hi Den, 

Can you remember what version was installed before v1.11.1?

Knowing this will help me narrow down any changes that may have caused your issues.

I suspect it was something earlier than V1.11.0 because the only change between 1.11.0 ~ 1.11.1 was to remove a debugging comment.

Thanks, Steve.

**Comment by Steve-Mcl** [comment]
Summary.

The `user` object is empty.

Rob also witnessed this when preparing for a customer demo and needed a working dashboard with the user object populated - through trial and error we found a work around as follows:
1. create new (fresh) instance of NR 3.1.9
2. install the plugin
3. restart instance
4. install dashboard
5. switch on security
6. restart instance

What we dont know: was using NR 3.1.9 the fix? was the order of events important? Were the restart(s) important?

**Comment by joepavitt** [comment]
If the `msg._client.user` object is present, but empty, then that means the `req.session.user` object is empty. The only place the `user` value is added is from within the plugin. So, the plugin is correctly being detected and run.

**Comment by joepavitt** [comment]
In the plugins `onSetup` callback, we run:

```js
socketio: {
    ...
    auth: {
        user: req.session?.user
    }
}
```

Then, when a `msg` is passed from the client to the server, we append the `_client` object with:

```js
msg._client = { ...msg._client, ...{ user: conn.handshake.auth.user } }
```

So, the `conn.handshake.auth.user` object is coming back empty, which originates from `req.session.user`

**Comment by joepavitt** [comment]
Okay, on an old instance, I can get it to work by "Unregister" on the PWA service worker:

<img width="750" alt="Screenshot 2024-06-11 at 10 10 18" src="https://github.com/FlowFuse/node-red-dashboard/assets/99246719/2a5c4071-2a92-4623-b1d7-8d6b7f106632">

Which forces a new service worker to run, and assigns the `session` correctly... yet another problem being caused by PWA

**Comment by joepavitt** [comment]
There were a couple of other steps I'd tried in this, and have found another old instance, will verify there.

**Comment by joepavitt** [comment]
This, very annoyingly, works exactly as I would expect it when I use this exact flow locally, which differs from what you're showing me in the screen recording. Trying to assess what the differences are.

**Comment by joepavitt** [comment]
Uploading Screen Recording 2024-03-08 at 14.49.36.mov…

One thing I do notice is that onblur of the text input re-sends the input, which can give the illusion of it being shared across the template as we don't have client-constraints active on the `text-input` field, but it's clear in your video recording the `template` updates _immediately_ in the other session.

I've tried with partial deploys, and full deploys, and I cannot reproduce the problem - it is very very odd

**Comment by SynoUser-NL** [comment]
Thanks for diving into this!

I am curious why you're mentioning "we don't have client-constraints active on the text-input field". Is this not yet implemented for that node?

As long as this doesn't happen between FlowFuse user sessions I guess we could live with the session bleed for now. Although I can't escape the feeling something is off with our Dashboard or configuration. And the fact that you cannot replicate what we're seeing only enhances that.

**Comment by SynoUser-NL** [comment]
Add to that another issue we've had for a couple of versions (but haven't gotten around to reporting) that may or may not be related: we're using several notification popups, about half of them set to display Center on screen, the others are positioned Top Center. Every one of the Center ones, on all dashboards, appear on the bottom center of the screen.

We are using a little CSS code for styling, nothing too fancy, and disabling that makes no difference. I don't think I've seen anyone else complain about this either, although I must admit I haven't been following the new issues as closely as I perhaps should. Or there are few people using the Center setting.
When I set the popup to appear Bottom Center, it appears bottom right. The rest all appears as configured.
And as I said we've seen this for a couple of versions now (somewhere from V1.08 onwards) and I think such a very visual anomaly would be addressed within a version or 2 at most. 
If this isn't an already known issue and you can replicate it, I'd be happy to submit an issue for it of course.

**Comment by SynoUser-NL** [comment]
So.. are we or have we been doing something that messes up our  Dashboard 2 configs in multiple ways?
How do we find out what without rebuilding the entire environment?

**Comment by colinl** [comment]
> Every one of the Center ones, on all dashboards, appear on the bottom center of the screen.

I am seeing that too (just tried it).

> When I set the popup to appear Bottom Center, it appears bottom right

That too.

**Comment by SynoUser-NL** [comment]
> > Every one of the Center ones, on all dashboards, appear on the bottom center of the screen.
> 
> I am seeing that too (just tried it).
> 
> > When I set the popup to appear Bottom Center, it appears bottom right
> 
> That too.

Glad I'm not the only one..
Thanks for checking!

**Comment by joepavitt** [comment]
> If this isn't an already known issue and you can replicate it, I'd be happy to submit an issue for it of course.

Yes please, we try not to pollute separate issues with different threads - new issues are _always_ welcome. I'd rather have duplicates, than it be missed entirely

**Comment by joepavitt** [comment]
> I am curious why you're mentioning "we don't have client-constraints active on the text-input field". Is this not yet implemented for that node?

I was just referring to the option not being ticked on the "Client Data" tab for Dashboard in the right-side menu

**Comment by SynoUser-NL** [comment]
> Yes please, we try not to pollute separate issues with different threads - new issues are _always_ welcome. I'd rather have duplicates, than it be missed entirely
> 
> Update: I've created the issue here: #960

Thanks Joe!



> I was just referring to the option not being ticked on the "Client Data" tab for Dashboard in the right-side menu

Ok. but even ticket this doesn't make a difference.

So how do we proceed from here in resolving the issue? Is there anything we can do to provide more data \ logging ?

**Comment by joepavitt** [comment]
> So how do we proceed from here in resolving the issue? Is there anything we can do to provide more data \ logging ?

A call would be useful I think, can screenshare, iterate on thoughts/ideas better. I'll e-mail you a link to my Calendly

**Comment by joepavitt** [comment]
On a call with @SynoUser-NL, we come to the following conclusions:

- The issue being seen is when the _same_ user is signed in to different browsers. This is because our logic currently treats the `msg._client.user` as an override of the `socketId` check. It doesn't care if the `socketId` is different, as long as the `user` object matches. We did this because we wanted to account for browser refreshes, where the auth session is persistent, but the `socketId` is not.
- Discussing the use case with Den, I agree that the `socketId` should be considered here. It would behave more how we expect a "traditional" web application too.
- The use case Iw as concerned about is a browser refresh, or if a user _only_ knows the user details on the server-sixde they want to interface with, and not the `socketId`.

Actions to take:
- Modify the connection checks to account for _both_ `socketId` (if present) _and_ `_client.user...` (if present). Checks should _all_ pass, not just considering a single pass (when matching with user) as success.

Fairly sure the logic here:

https://github.com/FlowFuse/node-red-dashboard/blob/b5255b7d999eff181ef98e507c9fcea564da071f/nodes/store/data.js#L14-L36

is what needs to be updated to reflect the above changes.
